lib/rails_pg_adapter/patch.rb

```diff
@@ -43,43 +43,41 @@ def missing_column_error?(error_message)
 
     def exec_cache(*args)
       sleep_times = ::RailsPgAdapter.configuration.reconnect_with_backoff.dup
+      query, _ = args
       begin
         super(*args)
       rescue ::ActiveRecord::StatementInvalid,
              ::ActiveRecord::ConnectionNotEstablished,
              ::ActiveRecord::NoDatabaseError => e
         raise unless ::RailsPgAdapter::Patch.supported_errors?(e)
 
-        if try_reconnect?(e)
-          sleep_time = sleep_times.shift
-          handle_error(e) unless sleep_time
-          warn("Retry query failed, retrying again in #{sleep_time} sec.")
-          sleep(sleep_time)
-          retry
-        else
-          handle_error(e)
-        end
+        handle_error(e) || raise unless try_reconnect?(e)
+        sleep_time = sleep_times.shift
+        handle_error(e) || raise unless sleep_time
+        warn("Retry query failed, retrying again in #{sleep_time} sec. Retrying: #{query}")
+        sleep(sleep_time)
+        connect
```

Comment:
this is the line being added mostly, rest is just moving the exception handling.